allennlp/training/trainer_base.py

```diff
@@ -0,0 +1,111 @@
+"""
+A :class:`~allennlp.training.trainer.Trainer` is responsible for training a
+:class:`~allennlp.models.model.Model`.
+
+Typically you might create a configuration file specifying the model and
+training parameters and then use :mod:`~allennlp.commands.train`
+rather than instantiating a ``Trainer`` yourself.
+"""
+# pylint: disable=too-many-lines
+
+import logging
+from typing import Dict, List, Union, Any
+
+import torch
+import torch.optim.lr_scheduler
+
+from allennlp.common import Params, Registrable
+from allennlp.common.checks import ConfigurationError, check_for_gpu
+from allennlp.models.model import Model
+from allennlp.nn import util as nn_util
+from allennlp.training import util as training_util
+
+logger = logging.getLogger(__name__)
+
+
+class TrainerBase(Registrable):
+    """
+    The base class for an AllenNLP trainer. It can do pretty much
+    anything you want. Your subclass should implement ``train``
+    and also probably ``from_params``.
+    """
+    default_implementation = "default"
+
+    def __init__(self,
+                 model: Model,
```

Comment:
As you noted, having a single model here precludes fancy stuff with GANs, etc.  If we have this separate base class, we should probably just not include the model.  `batch_loss` could be a utility function that takes the `Model` as an argument, and the only other thing this does is move the model to the GPU, which could just be done in the concrete class's constructor (or in a utility method called by the constructor).